### ISSUE TITLE
For exists check, text in the element should be considered and not UI elements in the same level

### ISSUE THREAD

**Issue by sguptatw** [getgauge/taiko#217]
**Expected behavior**
For exists check, text in the element should be considered and not UI elements in the same level.

**Actual behavior**
`text(contains(<text>))` works but not `text(<text>)`. Here the parent div contains the `<text> <comboBox>`

**Steps to replicate**
```
const { openBrowser, goto, toRightOf, checkBox } = require('taiko');

(async () => {
    try {
        await openBrowser({headless:false});
        await goto("https://react-shopping-cart-67954.firebaseapp.com/");
        console.log(await text("Order by").exists())
    } catch (e) {
        console.error(e);
    } finally {
        closeBrowser();
    }
})();
```
> The console log prints `Does not exist` where as `await text(contains("Order by"))` prints `Exists`. There is no other text in the parent element. However there is another comboBox along with this text.

**Version**
```
commit - 09afc93da99da144aa002a91feb1bd4b44506010
```

**Comment by negiDharmendra** [comment]
@sguptatw I wonder what would be the use case where we want the [text](https://taiko.gauge.org/#text)  API to behave like [contains](https://taiko.gauge.org/#contains) API.

**Comment by NivedhaSenthil** [comment]
Adding to @negiDharmendra  currently the default behaviour of `text` api is to search for exact text across element (i.e) textContent of the UI element not the innerText. The textContent of the div in the above case is "Order by SelectLowest to highestHighest to lowest".  If only the text on element to be considered the text spanning across element or having `<br>` in between will not be searchable using `text` api

**Comment by sguptatw** [comment]
@negiDharmendra @NivedhaSenthil 
From a user's perspective this behavior is not ideal

The problem is that when the user looks at the screen one will look for console.log(await text("Order by").exists()) and expect it to Exist. One will not give the values of the combo box along with text "Order by" for checking if it exists.

![orderby](https://user-images.githubusercontent.com/6310197/47557544-9701ca00-d92e-11e8-9a44-459795c91e5a.png)

**Comment by negiDharmendra** [comment]
@sguptatw Totaly agree on the point that user would expect `await text("Order by").exists()` to say that the element does exist. 
My only concern here is that taiko already has an [API](https://taiko-preview.gauge.org/#contains) which full fill such requirement. It might be confusing for the user to decide which one to use when if both of them ( [text](https://taiko-preview.gauge.org/#text) and [contains](https://taiko-preview.gauge.org/#contains)) behave similarly.

**Comment by sguptatw** [comment]
@negiDharmendra If user explicitly mentions text then I am looking for an exact match. Until I specify contains. eg:- text(contains("Order by")) would match text with `sdfsdf Order by sdfsdf`. However text("Order by") should show exists if it is the exact match isn't it?

**Comment by sguptatw** [comment]
**Discussion pending**
From a user's perspective, the expectation here is that `comboBox(near(text("Order by"))).select("Lowest to highest")` should work.

**Comment by zabil** [comment]
> From a user's perspective, the expectation here is that comboBox(near(text("Order by"))).select("Lowest to highest") should work.

Agree with this. To the user "Order by" is the text on the UI, so the `text("Order by")` must work.

**Comment by sriv** [comment]
Note that:

- in this example, for `text("Order by")` to work,i.e locate the element, it needs to do a `contains` match. By design `text` does an exact match, so by changing `text` api, it will be synonymous to `contains`

- there is another option, to omit text of elements that have an independent API, i.e. in this case there is a `combobox` api, so `text` can omit all combobox text while matching. This will however break in at least the below cases:

  - if the combobox is disabled, then it might make sense to get the complete text ex. `Order By Lowest Price`
  -  If the elements look visually like a combobox or a checkbox, but semantic html is not followed. ex. use of an image to indicate a checkbox rather than an actual `input` element.

**Comment by sguptatw** [comment]
@sriv Thanks. That summarizes yesterday's discussion. Here is another approach I was thinking about
 
- First look if the text matches exactly for a given element if yes return the text element else look across html elements to perform the operation. 
Yes this can take longer for elements spanning across multiple elements, 
> Some performance optimization can be done with proximity selectors and can be addressed based on any issues that we encounter in future.

Currently the text API looks for an exact match. This will continue to do that, without needing to white-list or black-list of elements to be considered for text input.

Thoughts?

**Comment by sriv** [comment]
> First look if the text matches exactly for a given element if yes return the text element else look across html elements to perform the operation.

How would this be different from `contains` ?

**Comment by NivedhaSenthil** [comment]
Even the above suggested approach wont help for this issue as there is no UI element only with the text `Order by`.

**Comment by sguptatw** [comment]
**Discussion Update**
As suggested earlier, if text API does contains by default then this use case gets addressed. Another concern was how should exact match be handled?
- no exact match until there is a use case
- The text API can take a parameter to indicate if exactMatch is required
`text(<text>,{exactMatch:true})` this can be false by default

Thoughts?

**Comment by NivedhaSenthil** [comment]
Linking the issue where `text` api was changed to exact match for more reference https://github.com/getgauge/taiko/issues/40

**Comment by sguptatw** [comment]
**Discussion update**
- contains API will be deprecated
- text API will do an exactMatch of text first, if it is not available then it will do a  contains find to match the specified element with text.

**Comment by NivedhaSenthil** [comment]
Fix done in dc9856c

**Comment by sguptatw** [comment]
So, now `console.log(await text("Order by SelectLowest to highestHighest to lowest").exists())` and `console.log(await text("Order by").exists())` is `Exists`
